File: gfapy/line/edge/gfa1/references.py
import gfapy

class References:
  def _initialize_references(self):
    for d in ["from_segment", "to_segment"]:
      s = self._gfa.segment(self.get(d))
      if s is None:
        if self._gfa._segments_first_order:
          raise gfapy.NotFoundError()
        s = gfapy.line.segment.GFA1({"name" : self.get(d),
                                     "sequence" : "*"},
                                    version = "gfa1",
                                    virtual = True)
        s.connect(self._gfa)
      self._set_existing_field(d, s, set_reference = True)
      if self.record_type == "L":
        et = self.from_end.end_type \
            if d == "from_segment" else self.to_end.end_type
        key = "dovetails_{}".format(et)
      else:
        key = "edges_to_contained" if (d == "from_segment") \
              else "edges_to_containers"
      s._add_reference(self, key)

  def _import_field_references(self, previous):
    for d in ["from_segment", "to_segment"]:
      self._set_existing_field(d, self._gfa.segment(self.get(d)),
          set_reference = True)

  def _backreference_keys(self, ref, key_in_ref):
    if ref.record_type == "P":
      return ["paths"]
    elif ref.record_type == "S":
      return ["from_segment", "to_segment"]
    else:
      return []

File: lib/rgfa/line/edge/gfa1/references.rb
RGFA::Line::Edge::GFA1 ||= Module.new

module RGFA::Line::Edge::GFA1::References

  private

  def initialize_references
    [:from, :to].each do |dir|
      s = @rgfa.segment(get(:"#{dir}_segment"))
      if s.nil?
        raise RGFA::NotFoundError if @rgfa.segments_first_order
        s = RGFA::Line::Segment::GFA1.new({:name => get(dir),
                                         :sequence => "*"},
                                         version: :gfa1,
                                         virtual: true)
        s.connect(@rgfa)
      end
      set_existing_field(:"#{dir}_segment", s, set_reference: true)
      if self.record_type == :L
        et = send(:"#{dir}_end").end_type
        key = :"dovetails_#{et}"
      else
        key = (dir == :from) ?
          :edges_to_contained :
          :edges_to_containers
      end
      s.add_reference(self, key)
    end
  end

  def import_field_references(previous)
    [:from_segment, :to_segment].each do |dir|
      set_existing_field(dir, @rgfa.segment(get(dir)), set_reference: true)
    end
  end

  def backreference_keys(ref, key_in_ref)
    case ref.record_type
    when :P
      [:paths]
    when :S
      [:from_segment, :to_segment]
    else
      []
    end
  end

end
